src/datachain/client/gcs.py

```diff
@@ -32,6 +32,15 @@ def create_fs(cls, **kwargs) -> GCSFileSystem:
 
         return cast(GCSFileSystem, super().create_fs(**kwargs))
 
+    def url(self, path: str, expires: int = 3600, **kwargs) -> str:
+        try:
+            return self.fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
+        except AttributeError as exc:
+            is_anon = self.fs.storage_options.get("token") == "anon"
+            if is_anon and "you need a private key to sign credentials" in str(exc):
```

Comment:
a few questions come to my mind: :)

can we return a public link right away if anon is set to True?

is there any guarantee that HTTP link this always work for public buckets?

do we do anything like this for S3? if not, why?

when does this error happen?

Reply:
> a few questions come to my mind: :)

These are good questions! 👍

> can we return a public link right away if anon is set to True?

Yes, we can. I doubted about chosen option [here](https://github.com/iterative/datachain/pull/748#discussion_r1897985463) and still not sure, but on the other side I just wanted to be sure I will not break any existing use case.

> is there any guarantee that HTTP link this always work for public buckets?

Yes: https://cloud.google.com/storage/docs/access-public-data#api-link

> do we do anything like this for S3? if not, why?

We definitely should do the same for other types of storages. I am not doing it here for some reasons:

- I wanted to check first if this will solve all the issues in SaaS
- it will requires more work and proper tests, I am not sure we have capacity now for this (fix for GCP was quick and simple)

> when does this error happen?

Original issue is here: https://github.com/iterative/studio/issues/11118

Reply:
> Original issue is here: https://github.com/iterative/studio/issues/11118

yep, I understand. I guess my question is why it is happening. If anon it passed AFAIU it should not be even trying to use credentials (I assume it is failing on a cluster because some default Google cred are present, right?)

Reply:
> yep, I understand. I guess my question is why it is happening. If anon it passed AFAIU it should not be even trying to use credentials (I assume it is failing on a cluster because some default Google cred are present, right?)

Ah, got it. It is failing not in a workers cluster, but on the backend (API request). It is interesting, I can not find any default Google credentials on backend 🤔